### PR TITLE
Allow inference configs to not rely on training and validation

### PR DESCRIPTION
**Pull request by yahskapar**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ def signal_method_inference(config, data_loader):
     print(config, end='\n\n')
 
     data_loader_dict = dict()
-    if config.TOOLBOX_MODE == "train_and_test" or config.TOOLBOX_MODE == "only_test":
+    if (config.TOOLBOX_MODE == "train_and_test" or config.TOOLBOX_MODE == "only_test"):
```
Comment (xliucs) [inline main.py:127]: I dont think you need parentheses

File: main.py [hunk 2]
```diff
@@ -154,7 +154,7 @@ def signal_method_inference(config, data_loader):
         else:
             raise ValueError("Unsupported dataset! Currently supporting UBFC, PURE, and SCAMPS.")
 
-        if config.TEST.USE_LAST_EPOCH:
+        if (config.TOOLBOX_MODE == "train_and_test" and config.TEST.USE_LAST_EPOCH):
```
Comment (xliucs) [inline main.py:157]: same. no parentheses

File: main.py [hunk 3]
```diff
@@ -170,7 +170,8 @@ def signal_method_inference(config, data_loader):
         else:
             raise ValueError("Unsupported dataset! Currently supporting UBFC, PURE, and SCAMPS.")
 
-        if config.TRAIN.DATA.DATASET is not None and config.TRAIN.DATA.DATA_PATH:
+        if (config.TOOLBOX_MODE == "train_and_test" and config.TRAIN.DATA.DATASET is not None and
```
Comment (xliucs) [inline main.py:173]: same. Can you also replace all `==` to `is` ?

Comment (xliucs) [reply main.py:173]: consider adding a comment to explain what does this if statement mean.

Comment (xliucs) [reply main.py:173]: Why not `config.TOOLBOX_MODE is "train_and_test" and config.TRAIN.DATA.DATASET and config.TRAIN.DATA.DATA_PATH`?

File: main.py [hunk 4]
```diff
@@ -186,7 +187,8 @@ def signal_method_inference(config, data_loader):
         else:
             data_loader_dict['train'] = None
 
-        if config.VALID.DATA.DATASET is not None and config.VALID.DATA.DATA_PATH and not config.TEST.USE_LAST_EPOCH:
+        if (config.TOOLBOX_MODE == "train_and_test" and config.VALID.DATA.DATASET is not None and 
+            config.VALID.DATA.DATA_PATH and not config.TEST.USE_LAST_EPOCH):
```
Comment (xliucs) [inline main.py:191]: same as above.

File: main.py [hunk 5]
```diff
@@ -202,7 +204,7 @@ def signal_method_inference(config, data_loader):
         else:
             data_loader_dict['valid'] = None
 
-        if config.TEST.DATA.DATASET is not None and config.TEST.DATA.DATA_PATH:
+        if (config.TEST.DATA.DATASET is not None and config.TEST.DATA.DATA_PATH):
```
Comment (xliucs) [inline main.py:207]: remove ()

File: neural_methods/trainer/DeepPhysTrainer.py [hunk 6]
```diff
@@ -32,12 +31,15 @@ def __init__(self, config, data_loader):
...
         self.model = DeepPhys(img_size=config.TRAIN.DATA.PREPROCESS.H).to(self.device)
         self.model = torch.nn.DataParallel(self.model, device_ids=list(range(config.NUM_OF_GPU_TRAIN)))
-        self.criterion = torch.nn.MSELoss()
-        self.optimizer = optim.AdamW(
-            self.model.parameters(), lr=config.TRAIN.LR, weight_decay=0)
-        # See more details on the OneCycleLR scheduler here: https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.OneCycleLR.html
-        self.scheduler = torch.optim.lr_scheduler.OneCycleLR(
-            self.optimizer, max_lr=config.TRAIN.LR, epochs=config.TRAIN.EPOCHS, steps_per_epoch=self.num_train_batches)
+        
+        if (config.TOOLBOX_MODE == "train_and_test"):
```
Comment (xliucs) [inline neural_methods/trainer/DeepPhysTrainer.py:35]: remove ()

**Review by xliucs — COMMENTED** [review]
I think you also need to update configs files under `configs/inference_configs` just remove train/valid blocks.

File: main.py (hunk 3 above)
Comment (yahskapar) [reply main.py:173]: > same. Can you also replace all `==` to `is` ?

`==` here makes sense because of the string literal comparison versus an identity comparison.

> Why not `config.TOOLBOX_MODE is "train_and_test" and config.TRAIN.DATA.DATASET and config.TRAIN.DATA.DATA_PATH`?

Fair point here since config.TRAIN.DATA.DATASET being blank (None type) here would also return False as expected. I'll update the commit in this PR with some comments.

Comment (yahskapar) [reply main.py:173]: Also, for a few of the multi-line conditional statements, I've kept parentheses in order to not exceed the maximum line length of 80 characters by using [implicit line joining](https://docs.python.org/3/reference/lexical_analysis.html#implicit-line-joining).

**Comment by yahskapar** [comment]
Tested with latest update to this PR using UBFC pre-trained model on PURE (/configs/infer_configs/UBFC_PURE_TSCAN_BASIC.yaml):

===Testing===                                                                                                                        
Testing uses pretrained model!                                                                                                       
                                                                                                                                     
FFT MAE (FFT Label):3.7277747844827585                                                                                               
FFT RMSE (FFT Label):13.965800381125103                                                                                              
FFT MAPE (FFT Label):3.4018523802081106
FFT Pearson (FFT Label):0.8179536971055215

File: evaluation/metrics.py [hunk 7]
```diff
@@ -45,10 +45,10 @@ def calculate_metrics(predictions, labels, config):
         prediction = _reform_data_from_dict(predictions[index])
         label = _reform_data_from_dict(labels[index])
 
-        if config.TRAIN.DATA.PREPROCESS.LABEL_TYPE == "Standardized" or \
-                config.TRAIN.DATA.PREPROCESS.LABEL_TYPE == "Raw":
+        if config.TEST.DATA.PREPROCESS.LABEL_TYPE == "Standardized" or \
+                config.TEST.DATA.PREPROCESS.LABEL_TYPE == "Raw":
             diff_flag_test = False
-        elif config.TRAIN.DATA.PREPROCESS.LABEL_TYPE == "DiffNormalized":
+        elif config.TEST.DATA.PREPROCESS.LABEL_TYPE == "DiffNormalized":
```
Comment (xliucs) [inline evaluation/metrics.py:51]: Are you sure if it is gonna work?  Have you tried a training session? Why only `config_TEST`

File: main.py [hunk 8]
```diff
@@ -127,7 +127,9 @@ def unsupervised_method_inference(config, data_loader):
     if config.TOOLBOX_MODE == "train_and_test" or config.TOOLBOX_MODE == "only_test":
         # neural method dataloader
         # train_loader
-        if config.TRAIN.DATA.DATASET == "COHFACE":
+        if config.TOOLBOX_MODE == "only_test":
+            pass
+        elif config.TRAIN.DATA.DATASET == "COHFACE":
             # train_loader = data_loader.COHFACELoader.COHFACELoader
```
Comment (xliucs) [inline main.py:133]: I am very confused about the logic. 

```
    if config.TOOLBOX_MODE == "train_and_test" or config.TOOLBOX_MODE == "only_test":
        # neural method dataloader
        # train_loader
        if config.TOOLBOX_MODE == "only_test":
            pass
```

Comment (xliucs) [reply main.py:133]: This does not make sense at all.

File: main.py [hunk 9]
```diff
@@ -140,7 +142,9 @@ def unsupervised_method_inference(config, data_loader):
             raise ValueError("Unsupported dataset! Currently supporting UBFC, PURE, and SCAMPS.")
 
         # valid_loader
-        if config.VALID.DATA.DATASET == "COHFACE":
+        if config.TOOLBOX_MODE == "only_test":
+            pass
+        elif config.VALID.DATA.DATASET == "COHFACE":
```
Comment (xliucs) [inline main.py:147]: Same as above. I think we need to think about a better way to address it. Maybe have a big if statement at outside. 

```
if config.TOOLBOX_MODE == 'train_and_test':
   DO SOMETHING
elif config.TOOLBOX_MODE == 'only_test':
  DO SOMETHING
 else:
  raise ValueError('Unsupported')
 ```

**Review by xliucs — COMMENTED** [review]
We should re-think how to structure main.py now. Otherwise, this PR looks good.

File: evaluation/metrics.py (hunk 7 above)
Comment (yahskapar) [reply evaluation/metrics.py:51]: Yes, I've tested this with a training session (both with a few train configs and infer configs) and didn't notice any issue. Only `config.TEST` matters in this case because the `calculate_metrics()` function, in `metrics.py`, is only ever called when we call the `test()` function for any trainer (TS-CAN, DeepPhys, PhysNet, EfficientPhys). Therefore, it shouldn'y matter what the other config (e.g., train or valid) parameters are.

File: main.py (hunk 9 above)
Comment (yahskapar) [reply main.py]: Take a look at 201a317 that just got added to this PR, hopefully the logic makes sense now.

File: main.py (hunk 8 above)
Comment (yahskapar) [reply main.py]: Take a look at 201a317 that just got added to this PR, hopefully the logic makes sense now. @xliucs

**Review by xliucs — APPROVED** [review]
LGTM!
